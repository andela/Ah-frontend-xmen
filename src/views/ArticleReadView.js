import React from 'react';
import { connect } from 'react-redux';
import { ToastContainer } from 'react-toastify';
import ArticlComponent from '../components/Articles/ArticleComponent';
import getSingleArticle from '../actions/articleActions/getSingleArticle';
import CommentView from './commentView/commentView';


function parseShareLinks(article, slug) {
  const articleUrl = `${process.env.FRONTEND_BASE_URL}/articles/${slug}`;
  const shareBody = `This is an interesting article, ${article.title} ${articleUrl}`;
  const mailshare = `mailto:?Subject=${encodeURI(article.title)}&body=${shareBody}`;
  const fbshare = `https://www.facebook.com/sharer/sharer.php?u=${encodeURI(articleUrl)}`;
  const gpshare = `https://plus.google.com/share?url=${encodeURI(articleUrl)}`;
  const twshare = `https://twitter.com/home?status=${encodeURI(shareBody)}`;

  return {
    mailshare, fbshare, gpshare, twshare,
  };
}


export class ArticleReadView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      notFound: false,
      loading: true,

    };
  }


  componentWillMount() {
    const { props } = this;
    const { slug } = props.match.params;
    props.getSingleArticle(slug);
  }

  componentWillReceiveProps(props) {
    this.setState({
      loading: false,
    });
  }

  render() {
    const { article } = this.props;
    if (this.state.loading) {
      return (
        <div className="h-100 w-100 p-5 m-5  text-center">
          <span style={{ fontSize: '400%' }}>Loading.....</span>
        </div>
      );
    }

    if (article.title === '') {
      return (
        <div className="h-100 w-100 p-5 m-5  text-center">
          <span style={{ fontSize: '400%' }}>404</span>
        </div>
      );
    }

    article.share_links = parseShareLinks(article, this.props.match.params.slug);

    return (
      <div>
        <ArticlComponent
          slug={this.props.match.params.slug}
          {...article}
        />
        <CommentView />
        <ToastContainer />
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  article: state.articleReducer.article,
  loading: state.articleReducer.loading,
});
export default connect(mapStateToProps, { getSingleArticle })(ArticleReadView);
