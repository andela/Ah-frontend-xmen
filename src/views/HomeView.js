import React from 'react';
import App from '../components/LandingPage/Banner';
import Articles from '../components/LandingPage/Articles';

const HomeView = () => (
  <div>
    <App />
    <Articles />
  </div>
);

export default HomeView;
