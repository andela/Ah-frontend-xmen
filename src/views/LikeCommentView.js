import React from 'react';
import { connect } from 'react-redux';
import likeCommentAction from '../actions/commentActions/commentActions';
import CommentBanner from '../components/comments/commentBanner';


class LikeCommentView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLiked: false,
    };
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick=(e) =>{
    console.log('hey');
  }

  render() {
    return (
      <CommentBanner onClick={this.handleClick} />
    );
  }
}
const mapStateToProps = state => ({
  likeCommentsAction: state.likeCommentReducer,
});
export default connect(mapStateToProps, { likeCommentAction })(LikeCommentView);
