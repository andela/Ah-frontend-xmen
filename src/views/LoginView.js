import React from 'react';
import Login from '../components/Login';
import SocialAuthView from './socialAuthView/socialAuthView';

const LoginView = props => (
  <div>
    <SocialAuthView />
    <Login />

  </div>
);

export default LoginView;
