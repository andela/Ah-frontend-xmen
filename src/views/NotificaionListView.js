
import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import notificationsAction from '../actions/notificationsAction';


export class NotificationListView extends React.Component {
  render() {
    const notifications = this.props.notifications.map((notification, index) => {
      const iconClass = notification.isUnread ? 'fas fa-circle m-3' : 'far fa-circle m-3';

      return (
        <Link to={notification.actionLink}>
          <li className="list-group-item " key={index}>
            <i className={iconClass} />
            {notification.message}
          </li>
        </Link>

      );
    });

    return (
      <div className="note-list container p-3">
        <div className="display-3">
                All notifications
        </div>
        <hr />
        <ul>
          {notifications}
        </ul>
      </div>
    );
  }
}


export function mapStateToProps(state) {
  return {
    auth: state.Auth,
    notifications: state.notificationsReducer.notifications,
    hasUnread: state.notificationsReducer.hasUnread,
  };
}

export default connect(
  mapStateToProps,
  {
    getNotifications: notificationsAction,
    markAsRead: notificationsAction,
  },
)(NotificationListView);
