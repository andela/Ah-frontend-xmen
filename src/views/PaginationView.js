/* eslint-disable react/no-unused-state */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import Pagination from '../components/Pagination/Pagination';
import fetchArticlesAction from '../actions/articleActions/fetchArticlesAction';

export class Paginations extends Component {
  constructor(props) {
    super(props);
    this.state = {
      pageNumbers: [],
      allArticles: 7,
      currentPage: 1,
      totalPages: 1,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.count && nextProps.count > 1) {
      if (nextProps.count > 6) {
        this.setState({
          allArticles: nextProps.count,
          totalPages: nextProps.count / 6,
        });
      } else {
        this.setState({ allArticles: nextProps.count, totalPages: 1 });
      }
    }
  }

  handleNext = (nextPageIndex) => {
    if (nextPageIndex < +this.state.totalPages && nextPageIndex >= 1) {
      this.props.fetchArticlesAction(nextPageIndex);
      this.setState({ currentPage: nextPageIndex });
    }
  };

  handlePrevious = () => {
    this.props.fetchArticlesAction(this.state.currentPage - 1);
  };

  render() {
    const { currentPage, totalPages } = this.state;
    let { pageNumbers } = this.state;
    pageNumbers = [];

    for (let i = 1; i <= totalPages; i += 1) {
      pageNumbers.push(i);
    }
    return (
      <Pagination
        pageNumbers={pageNumbers}
        nextPage={this.handleNext}
        previousPage={this.handlePrevious}
        currentPage={currentPage}
      />
    );
  }
}

export const mapStateToProps = state => ({
  newArticles: state.articles.articles,
  count: state.articles.allArticles,
});

export default connect(
  mapStateToProps,
  {
    fetchArticlesAction,
  },
)(Paginations);
