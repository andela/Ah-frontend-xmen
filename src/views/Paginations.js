import React, { Component } from 'react';
import { connect } from 'react-redux';
import Pagination from '../components/Pagination/Pagination';
import {
  fetchArticlesAction,
  getAllArticlesAction,
} from '../actions/articleActions/fetchArticlesAction';

class Paginations extends Component {
  constructor(props) {
    super(props);
    this.state = {
      pageNumbers: [],
      allArticles: 7,
      currentPage: 1,
    };
  }

  componentWillReceiveProps(nextProps) {
    console.log(nextProps);
    if (nextProps.count && nextProps.count > 1) {
      console.log(nextProps.count);

      this.setState({ allArticles: nextProps.count });
    }
  }

  handleNext = (nextPageIndex) => {
    this.props.getAllArticlesAction(nextPageIndex);
    this.setState({ currentPage: nextPageIndex });
  };

  handlePrevious = () => {
    this.props.getAllArticlesAction(this.state.currentPage - 1);
  };

  render() {
    const { allArticles, currentPage } = this.state;
    let { pageNumbers } = this.state;
    pageNumbers = [];
    let totalPages;
    if (allArticles > 6) {
      totalPages = allArticles / 6;
    }

    for (let i = 1; i <= totalPages; i++) {
      pageNumbers.push(i);
    }
    return (
      <Pagination
        pageNumbers={pageNumbers}
        nextPage={this.handleNext}
        previousPage={this.handlePrevious}
        currentPage={currentPage}
      />
    );
  }
}

export const mapStateToProps = state => ({
  newArticles: state.articles.articles,
  count: state.articles.allArticles,
});

export default connect(
  mapStateToProps,
  {
    fetchArticlesAction,
    getAllArticlesAction,
  },
)(Paginations);
