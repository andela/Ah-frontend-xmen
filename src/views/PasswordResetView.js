import React from 'react';
import { connect } from 'react-redux';
import { ToastContainer } from 'react-toastify';
import ResetForm from '../components/PasswordReset/ResetForm';
import passwordResetAction from '../actions/PasswordResetAction';
import passwordChangeAction from '../actions/PasswordChangeAction';
import NewPasswordForm from '../components/PasswordReset/NewPasswordFrom';
import Message from '../components/PasswordReset/Message';

export class PasswordResetView extends React.Component {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handlePasswordSubmit = this.handlePasswordSubmit.bind(this);
  }

    handleSubmit = (e) => {
      e.preventDefault();
      const email = e.target[0].value;
      this.props.passwordResetAction(email);

      this.props.passwordResetReducer.errors = undefined;
    }

    handlePasswordSubmit = (e) => {
      e.preventDefault();
      const password = e.target[0].value;
      const { token } = this.props.match.params;
      this.props.passwordChangeAction(password, token);
      this.props.passwordResetReducer.errors = undefined;
    }

    render() {
      if (this.props.passwordResetReducer.message !== undefined) {
        return (<div className="mt-5 pt-5 message-box "><Message message={this.props.passwordResetReducer.message} /></div>);
      }

      if (this.props.match.params.token === 'invalid-token') {
        return (<div className="mt-5 pt-5 message-box"><Message message="Sorry the link you followed is expired. Please request a new one." /></div>);
      }

      if (this.props.match.params.token !== undefined) {
        return (
          <div className="mt-5 mb-5  pt-5">
            <NewPasswordForm onSubmit={this.handlePasswordSubmit} />

          </div>
        );
      }

      return (
        <div className="mt-5 mb-5 pt-5">

          {/* <small>{this.props.passwordResetReducer.errors}</small> */}
          <ResetForm onSubmit={this.handleSubmit} />
          <ToastContainer />
        </div>
      );
    }
}

const mapStateToProps = state => ({
  ...state,
});

export default connect(mapStateToProps,
  {
    passwordResetAction,
    passwordChangeAction,
  })(PasswordResetView);
