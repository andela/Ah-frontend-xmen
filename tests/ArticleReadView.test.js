import React from 'react';
import { shallow } from 'enzyme';
import { toast } from 'react-toastify';
import { ArticleReadView, mapStateToProps } from '../src/views/ArticleReadView';


describe('ArticleReadView', () => {
  it('should render without crashing', () => {
    const props = {
      article: {
        title: 'test title',
        author: 'blash',
      },
      loading: false,
      match: {
        params: {
          slug: 'slug-as12',
        },
      },
      authState: {
        IsAuth: true,
      },
      getSingleArticle: jest.fn(),
      bookmarkListing: jest.fn(),
      bookmarkState: {
        successMessage: '',
        errorMessage: '',
      },
    };
    const instance = shallow(<ArticleReadView {...props} />);
    instance.setProps({
      loading: false,
    });
    expect(instance).toMatchSnapshot();
  });

  it('should render 404 view without crashing', () => {
    const props = {
      article: {
        title: '',
      },
      loading: true,
      match: {
        params: {
          slug: 'slug-as12',
        },
      },
      authState: {
        IsAuth: true,
      },
      getSingleArticle: jest.fn(),
      bookmarkListing: jest.fn(),
    };
    const instance = new ArticleReadView(props);
    instance.state = { notFound: true };
    expect(instance.render()).toMatchSnapshot();
  });


  it('should call setState after receiving props ', () => {
    const props = {
      bookmarks: [{ title: 'jj', slug: 'slug-as12' }],
      bookmarkState: {
        successMessage: '',
        errorMessage: '',
      },
      article: {
        title: 'test title',
        author: 'hello',
      },
      authState: {
        IsAuth: true,
      },
      bookmarkListing: jest.fn(),
      setState: jest.fn(),
      getSingleArticle: jest.fn(),
      componentWillReceiveProps: jest.fn(),
      match: {
        params: {
          slug: 'slug-as12',
        },
      },

    };
    const instance = shallow(<ArticleReadView {...props} />);
    instance.instance().setState = jest.fn();
    instance.instance().componentWillReceiveProps(props);
    expect(instance.instance().setState).toHaveBeenCalled();
  });

  it('should call getSingleArticle after will mount ', () => {
    const props = {
      articleReducer: {
        article: {
          title: 'test title',
        },
      },
      match: {
        params: {
          slug: 'slug-as12',
        },
      },
      getSingleArticle: jest.fn(),
      bookmarkListing: jest.fn(),
    };
    const instance = new ArticleReadView(props);
    instance.componentWillMount();
    expect(instance.props.getSingleArticle).toHaveBeenCalled();
  });
  it('should call onBookmark ', () => {
    const props = {
      articleReducer: {
        article: {
          title: 'test title',
        },
      },
      match: {
        params: {
          slug: 'slug-as12',
        },
      },
      authState: {
        IsAuth: true,
      },
      getSingleArticle: jest.fn(),
      bookmarkListing: jest.fn(),
      bookmarkArticle: jest.fn(),
    };
    toast.error = jest.fn();
    const instance = new ArticleReadView(props);
    instance.onBookmark({ preventDefault() {} });
    expect(instance.props.bookmarkArticle).toHaveBeenCalled();
  });
  it('should call handle flag ', () => {
    const props = {
      articleReducer: {
        article: {
          title: 'test title',
        },
      },
      match: {
        params: {
          slug: 'slug-as12',
        },
      },
      getSingleArticle: jest.fn(),
      bookmarkListing: jest.fn(),
    };
    toast.error = jest.fn();
    const instance = new ArticleReadView(props);
    instance.handleFlag();
    expect(toast.error).toHaveBeenCalled();
  });

  it('should call handle flag without token', () => {
    const props = {
      articleReducer: {
        article: {
          title: 'test title',
        },
      },
      match: {
        params: {
          slug: 'slug-as12',
        },
      },
      getSingleArticle: jest.fn(),
      bookmarkListing: jest.fn(),
    };
    localStorage.setItem('token', '123');
    window.location.assign = jest.fn();
    const instance = new ArticleReadView(props);
    instance.handleFlag();
    expect(window.location.assign).toHaveBeenCalled();
  });
  it('should map state to props', () => {
    const state = {
      articleReducer: {
        article: {},
        loading: true,
      },
      bookmarkListReducer: {
        bookmarks: [{ title: 'ded', slug: 'dededd' }],
      },
    };
    expect(mapStateToProps(state)).toEqual({
      article: {},
      loading: true,
      bookmarks: [{ title: 'ded', slug: 'dededd' }],
    });
  });

  it('should call getSingleAction after rating', () => {
    const props = {
      articleReducer: {
        article: {
          title: 'test title',
        },
      },
      match: {
        params: {
          slug: 'slug-as12',
        },
      },
      getSingleArticle: jest.fn(),
      ratingAction: jest.fn(),
    };
    global.setTimeout = jest.fn();
    jest.useFakeTimers();
    const instance = new ArticleReadView(props);
    instance.changeRating(5, props.slug);
    expect(global.setTimeout).toHaveBeenCalled();
    expect(global.setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 500);
    jest.runAllTimers();
  });

  it('should map state to props', () => {
    const state = {
      articleReducer: {
        article: {},
        loading: true,
      },
      bookmarkListReducer: {
        bookmarks: [],
        isBookmarked: undefined,
      },
    };
    expect(mapStateToProps(state)).toEqual({
      article: {},
      loading: true,
      bookmarks: [],
    });
  });
});
