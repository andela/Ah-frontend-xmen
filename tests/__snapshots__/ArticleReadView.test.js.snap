// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleReadView should render 404 view without crashing 1`] = `
<div
  className="h-100 w-100 p-5 m-5  text-center"
>
  <span
    style={
      Object {
        "fontSize": "800%",
      }
    }
  >
    404
  </span>
</div>
`;

exports[`ArticleReadView should render without crashing 1`] = `ShallowWrapper {}`;
